New specification (#349)
blockprotocol/blockprotocol: site/src/_pages/spec/2_graph-service-specification.mdx

- MUST be represented by an object, which:
  - MUST contain an `entityTypeId` identifying the type, which MUST be a str
  - MUST contain a `schema`, a JSON Schema object defining the constraints on the data in the entity’s `properties` field, which
  - MUST be of `type: "object"`
  - SHOULD use our JSON Schema extensions where appropriate
  - MAY otherwise contain any valid JSON Schema

i.e. entity types MUST conform to the following JSON schema:

```json
{
  "$id": "https://blockprotocol.org/types/services/graph/entity-type",
  "title": "EntityType",
  "type": "object",
  "properties": {
    "entityTypeId": {
      "type": "string"
    },
    "schema": {
      "$ref": "https://json-schema.org/draft/2020-12/schema"
    }
  },
  "required": ["entityTypeId", "schema"]
}
```

Blocks SHOULD use any schemas provided in the `entityTypes` message to understand the constraints on user input for each entity (e.g. what type(s) of data are valid for each property).

##### Links

Entities MAY be connected by a **link**, which:

- MUST contain:
  - `sourceEntityId`: the `entityId` of the source entity, which MUST be a string

>   - `destinationEntityId`– the id of a single entity the link is made to, which MUST be a string

Reviewer: ```suggestion
  - `destinationEntityId`: the id of a single entity the link is made to, which MUST be a string
```

  - MUST contain a `linkGroups` key, which contains a list of objects which group the links in the resolved graph (i.e. the edges) by entity and path. Each object:
    - MUST contain `links`, an array of links
    - MUST contain the `sourceEntityId`, identical across all entries in `links`
    - MUST contain the `path`, identical across all entries in `path`
  - MUST contain a `linkedEntities` key, which contains a list of the entities in the resolved graph (i.e. the nodes)
  - MUST contain a `depth` key, which indicates the depth to which the graph around the block entity has been resolved.
- Embedding applications MUST at least provide the links from the block entity. This represents a depth of `0`. They MAY also include the entities linked from the block entity and _their_ links which is a depth of `1`, and so on.
- See [blockGraph](#block-graph)

`linkGroups` should conform to the following schema:

```json
{
  "$id": "https://blockprotocol.org/types/services/graph/link-group",
  "title": "LinkGroup",
  "type": "object",
  "properties": {
    "links": {
      "type": "array",
      "items": {

>         "$ref": "https://blockprotocol.org/types/services/graph/link-group"

Reviewer: ```suggestion
        "$ref": "https://blockprotocol.org/types/services/graph/link"
```

      }
    },
    "sourceEntityId": {
      "type": "string"
    },
    "path": {
      "type": "string"
    }
  },
  "required": ["sourceEntityId", "destinationEntityId", "path"]
}
```

##### Linked aggregations

Blocks may also wish to link an entity to a particular aggregation of entities. To do so, blocks SHOULD create a **LinkedAggregationDefinition**, which:

- MUST contain:
  - `sourceEntityId`: the `entityId` of the source entity
  - `operation`: an aggregation operation which the embedding application should resolve the link to, following the structure of the `operation` object
  - `path`: the path to the field on the source entity this link is conceptually made on, expressed as a **[JSON path](https://goessner.net/articles/JsonPath/)**, which MUST be a string

i.e. a linked aggregation definition must conform to the following schema:

```json
{
  "$id": "https://blockprotocol.org/types/services/graph/linked-aggregation-definition",
  "title": "LinkedAggregationDefinition",
  "type": "object",
  "properties": {
    "sourceEntityId": {
      "type": "string"
    },
    "path": {
      "type": "string"
    },
    "operation": {

>       "type": "integer",
>       "inclusiveMinimum": 0

Reviewer: ```suggestion
      "$ref": "https://blockprotocol.org/types/services/graph/aggregation-operation"
```

    }
  },
  "required": ["sourceEntityId", "operation", "path"]
}
```

### Identification

#### <Hidden>Identification </Hidden>Summary

> Blocks must pass all of the **identifying fields** provided to them when making requests in respect of any entity.
> 
> If the embedding application provides a value for any identifying field, it should be assumed it is required to reliably identify the entity.

Reviewer: ```suggestion
Blocks must pass an `entityId` when making requests in respect of any entity. 

Embedding applications should ensure this is stable and unique for any given entity.
```

#### <Hidden>Identification </Hidden>Specification

> Blocks MUST provide all `entityId` when identifying a particular entity for the purpose of **messages** under the graph service.

Reviewer: ```suggestion
Blocks MUST provide `entityId` when identifying a particular entity for the purpose of **messages** under the graph service.
```

- `INVALID_INPUT`: the message relied on previous input that was missing or otherwise invalid

Individual messages MAY specify only a subset of these codes as being valid for that message.

### Message definitions

#### <Hidden>Message definitions </Hidden>Summary

Messages under the graph service can be grouped into two functional categories:

- **values** provided by the embedding application, marked as `source: "embedder"` and `sentOnInitialization: true`
- **requests** made by the block during its lifetime, marked as `source: "block"`

Note that there is no reason why embedding applications cannot make requests, nor why blocks cannot send values on initialization, but the service does yet not specify any such messages.

The **values** are all sent from the embedding application to the block on initialization, and re-sent subsequently if they change. They are:

- `blockEntity`: the block entity, conforming to the constraints defined in the block schema
- `entityTypes`: the entity type definitions for any entities provided to the block in other messages
- <span id="block-graph">`blockGraph`</span>: the entity graph with the block at

>   its root, which contains: - `linkedEntities`: the entities linked from the
>   block entity, and from those entities, and so on, resolved to a depth
>   determined by the embedding application - `linkGroups`: the links from the
>   block entity and the links from any entities in `linkedEntities`, grouped by
>   source entity and link path

Reviewer: ```suggestion
  its root, which contains: 
  - `linkedEntities`: the entities linked from the block entity, and from those entities, and so on, resolved to a depth determined by the embedding application 
  - `linkGroups`: the links from the block entity and the links from any entities in `linkedEntities`, grouped by source entity and link path
```

- `linkedAggregations`: field containing the definitions and results of aggregations linked from the block’s entity - these are special objects which contain an aggregation definition (e.g. in plain English, "top 10 Cars sorted by age, descending") and its results at the time of provision
- `readonly`: whether or not the block should display in ‘readonly’ mode, disabling or hiding any editing controls it offers.

The **requests** may be made by the block once it has been initialized, and represent operations which create, read, update or delete entities, entity types, and create links. They are named according to the pattern `[action][Resource]`, e.g. `createEntity`.

> 

Reviewer: ```suggestion

As well as the formal service JSON specification, the messages can be sent and received via the `@blockprotocol/graph` [package](https://www.npmjs.com/package/@blockprotocol/graph).

```

> #### <Hidden>Message definitions </Hidden>Specification
> 
> ##### Graph service JSON definition
> 
> See in the repo [here](https://github.com/blockprotocol/blockprotocol/blob/main/packages/%40blockprotocol/graph/src/graph-service.json#L5)

Reviewer: ```suggestion
#### <Hidden>Message definitions </Hidden>Specification

##### Graph service JSON definition

The messages available for exchange in the graph service are defined in the service's JSON definition [here](https://github.com/blockprotocol/blockprotocol/blob/main/packages/%40blockprotocol/graph/src/graph-service.json#L5)
```
